### ISSUE TITLE
sticky: Prefs vs Recommended Extensions: Co-Existance+Enhancement | Conflicts

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [arkenfox/user.js#350]
A list of *stuff* - thanks @smithfred for the ideas

Extensions can often work better than a global pref because of their flexibility (but may still have downsides or issues), or they can enhance/compliment them, or they can clash. Either way, there is always more than one way to skin a cat <sup>no cats were harmed in the making of this issue</sup>.

Here are some solutions that may fit your needs. If you have any other suggestions to this list-in-progress, please let us know :)

---
### :large_orange_diamond: Mixed Passive Content
* pref: `security.mixed_content.block_display_content`=true (`1241`)
* alternative:
   * set the pref to false
   * add `https-strict: * true` to uMatrix
   * allow exceptions on a site by site basis

---
### :large_orange_diamond: Web Workers
* pref(s): `dom.workers.enabled`=false (`2301`) and/or `dom.serviceworkers.enabled`=false (`2302`)
   * Note: `dom.workers.enabled` was [removed](https://bugzilla.mozilla.org/1434934) in FF60
* alternative:
   * set the pref(s) to true
   * add `no-workers: * true` to uMatrix
   * allow on a site by site basis
* note: 
   * FYI: uMatrix (and FF) can't use workers if cookies are set to "Keep until I close firefox" [1429714](https://bugzilla.mozilla.org/show_bug.cgi?id=1429714)
   * if FF prefs allow web workers, but uMatrix blocks them, there may be instances where the web site detects workers are available but can't actually use them, leading to breakage, rather than falling back to the non-web-worker version [read this somewhere on uM repo but can't find it now]

---
### :large_orange_diamond: Cookies [1]
* pref: `network.cookie.cookieBehavior`=2 (block all) (`2701`)
   * use FF site exceptions (`allow` eg for logins, `allow for session` eg for sites that require them to work)
* alternative:
   * set pref to allow cookies (recommended value of 1 for same host only)
   * add `* * cookie block` to uMatrix
   * allow on a granular level
* notes: [uMatrix wiki on cookies](https://github.com/gorhill/uMatrix/wiki/Cookies)
   * "Blacklisted cookies are not prevented by uMatrix from entering your browser. However they are prevented from leaving your browser"
   * This is about standard HTTP Cookie header. Cookies can still be read, and sent to remote servers by javascript in other data structures (POST request for example)
* issues:
   * cookies **control** other persistent storage (local storage, IndexedDB), both when allowing creation and for FF and extensions when cleaning. While uMatrix has a setting to automatically delete blocked cookies, currently FF/Extensions have issues/limitations with sanitizing/FPI. Just removing a cookie can leave behind orphaned persistent data <sup>99% sure AFAIK</sup>

---
### :large_orange_diamond: Cookies [2]
* pref: `network.cookie.cookieBehavior`=2 (block all) (`2701`)
   * use FF site exceptions (`allow` eg for logins, `allow for session` eg for sites that require them to work)
* alternatives
   * set pref to allow cookies (recommended value of 1 for same host only)
       * I *think* Cookie Autodelete *requires* cookies to be enabled or it won't work
   * cookie extensions
* notes
   * these block/allow cookies coming in. They allow lots of flexibility (white-black-greylists, auto-clearing on tab close etc), and will one day be able to properly handle related persistence storage and FPI (see issues for Cookies [1])

---
### :large_orange_diamond: Canvas
* pref(s): `privacy.resistFingerprinting`=true (`4501`) (RFP)
   * so much is bundled under this pref that it is not feasible to set it to false
   * for the gazillion things RFP does see #7 
* alternatives:
   * CanvasBlocker (or similar)
* notes:
   * RFP canvas protection currently prompts for every site (but can remember site permissions). In FF59+ canvas has been [added](https://bugzilla.mozilla.org/show_bug.cgi?id=1413780) to the PageInfo>Permissions panel. RFPCanvas [might](https://bugzilla.mozilla.org/show_bug.cgi?id=1429865) have a pref for its `default permission` (but looks unlikely). Hopefully it will also be added to the options Site Preferences section for site management. Update: preferred fix is to [restrict/lower](https://bugzilla.mozilla.org/show_bug.cgi?id=1376865) the prompts
* strategy:
   * set CanvasBlocker to fake. Do NOT set to block as this will disable the API and you will not get the same result as RFP.
   * block sites when prompted: RFP takes over and CB is never used
   * allow sites IF you must: RFP allows CB to take over which will fake
   * use a CB whitelist for sites that MUST have the real thing

---
### :large_orange_diamond: Referers
 * pref(s): section 1600
   * these are *global* settings, and at defaults and in our user.js, they only limit some data in some cases (otherwise way too much of the internet breaks)
   * if you want ANY real control over referers, you NEED an extension
 * terms:
   * source: the site you loaded, 1st party
   * destination: a site linked to, or the site being requested for 3rd party content
   * example: 
      * load `SiteA` (SOURCE, 1st party)
      * `SiteA` requests content from 3rd party `SiteB`, `SiteC`, `SiteD` (B,C,D are DESTINATION)
      * `SiteA` might include a referer (basically saying "Hi I'm requesting this for `SiteA`") to any or all of those destination sites
 * EXTRA Solution 1: uMatrix
   * spoof by default: `referrer-spoof: * true`
      * Dashboard>Settings>Privacy>"Spoof HTTP referrer string of third-party requests"
      * uMatrix spoofs any referers as the `destination` (see [this](https://github.com/gorhill/uMatrix/wiki/Per-scope-switches#spoof-referer-header) uMatrix wiki entry)
   * allow exceptions on a per scope basis (this is under the three vertical dots dropdown in your uMatrix panel)
   * uMatrix can only control the SOURCE, not per DESTINATION (so in the example above, Sites B C + D would all get referers)
 * EXTRA Solution 2: Smart Referer (instead of uMatrix's setting)
   * whitelist, blacklist, built-in whitelist to reduce breakage, etc
   * allows source-destination control: so in the example above, you could allow SiteB but still block Sites C + D
   * allows blocking, spoofing as destination, faking, allowing, etc
   * the ultimate referer tool

**Comment by earthlng** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1429714 is about **service** workers.
Since there are problems with the way extensions use Content-Security-Policies to block these things, I wouldn't rely on uMatrix too much. The prefs are the most reliable way to block stuff.

> uMatrix allows cookies IN but blocks them from going OUT

that's how it's supposed to work but extensions don't see the changes made to headers by other extensions and therefore it's possible that the cookie(s) will still be sent.

**Comment by Thorin-Oakenpants** [comment]
^^ feel free to correct the OP if I make factual errors

**Comment by theWalkingDuck** [comment]
> > uMatrix allows cookies IN but blocks them from going OUT

> that's how it's supposed to work but extensions don't see the changes made to headers by other extensions and therefore it's possible that the cookie(s) will still be sent.

Either you're using extensions which are adding there own cookies to the request header (after uMatrix striped it) or you're using malicious extensions which are sending a hidden/background request.

The question is: Why on earth are you using such extensions ?

**Comment by earthlng** [comment]
> after uMatrix striped it

all extensions receive the headers in parallel and unmodified. It doesn't matter if uMatrix striped it because the next extension still sees the original cookie header.

> The question is: Why on earth are you using such extensions ?

I don't but there are apparently people who have like 70+ addons or whatnot, and I'm just saying that it *can* happen. Here are just 2 examples of legitimate extensions that *do* or *could* modify the cookie header in certain situations:

https://github.com/snyderp/web-api-manager/blob/master/add-on/background_scripts/background.js#L58
https://github.com/scakemyer/cloudhole/blob/master/background.js#L216

**Comment by earthlng** [comment]
@Thorin-Oakenpants mixed-content + uMatrix is probably worth mentioning here, too.

**Comment by Thorin-Oakenpants** [comment]
@kkapsner In the first post I have a canvas section. Can you confirm the `strategy` part is correct, please.

See https://github.com/kkapsner/CanvasBlocker/issues/158#issuecomment-348899000 and your subsequent comment. I think I read you say recently that if FF's Canvas = allow then CB doesn't get a look in? Which is not what I expected with my old comment in that link

> I do not know what happens if you block (I assume CB is never used?)

If FF Canvas blocks = white square, right - or at least `undetermined` at panopticlick? And CB is never used, right?

> * I do not know what happens if you allow (I assume CB then takes over?)

This is the million dollar question. Rather than have me fluff around and get it wrong, I thought you might know the answer. Thanks

**Comment by kkapsner** [comment]
The strategy looks good.

If you block the canvas via FF the canvas is completely handled by it and CB never kicks in.
When you allow via FF the canvas is handled by CB which will do what ever you specified it to do.

**Comment by Thorin-Oakenpants** [comment]
updated the strategy. Looks like a default permission pref will not happen - in RFP its all or nothing, so they don't want to allow overrides to default behaviors. Instead [1376865](https://bugzilla.mozilla.org/show_bug.cgi?id=1376865) aims to reduce the prompts/fatigue.

However, regardless of prompts/default bahavior - every remembered canvas prompt => site data in persistent storage (note: not persistent in TBB) which sucks. Either way, I recommend RFP users set CB to block rather than fake by default to match

Q: does CB=block produce the same hash as RFPCanvas=deny? I think RFPCanvas just returns a white canvas (not up on any specs such as size etc)

**Comment by Thorin-Oakenpants** [comment]
> EDIT: when I messed around with this feature I think I also found a way to ignore the resistFingerprint setting (or answer). Not sure if this is useful or not...

Oooh - please reproduce this if you can (STR) - sounds like a bug

**Comment by kkapsner** [comment]
Block does block the API and therefore generates errors in most pages that just assume that the API is accessible. If you want the same behaviour than in  RFP=block you have to use the "white" random number generator.

The detection test is using canvas but only a minor subset of the API. The block mode may block before the RFP kicks in.

**Comment by Thorin-Oakenpants** [comment]
Ahhh .. makes sense re the API. Using block = makes RFP users more unique but gets rids of all [1] the prompts. Will test with white noise "random?"

[1] Not sure why reddit(?) and panopticlick (see #352 ) prompted when told to block (nothing in whitelist etc) - you may want to look into that

**Comment by Thorin-Oakenpants** [comment]
^^ ok, I made no sense whatsoever. In order to get a result the same as RFP, you have to use the CB fake setting which means RFP kicks in.

tl;dr: If you want the two to work together nicely and behave like RFP users, do not use block in CB, end of story :)

**Comment by smithfred** [comment]
Can't find where I mentioned the idea of general API blocking, but hey look:

https://www.ghacks.net/2018/01/30/webapi-manager-limit-website-access-to-web-apis/

**Comment by Atavic** [comment]
*3P Request Blocker* by [Sw](https://swclearnet.danwin1210.me/) prevents your browser from connecting to 3rd party resource without user consent. Only 61.0 KiB

**Comment by Atavic** [comment]
Re: https://github.com/pyllyukko/user.js/issues/348

**Comment by Thorin-Oakenpants** [comment]
@earthlng FYI: https://bugzilla.mozilla.org/show_bug.cgi?id=1439163

This would affect any extension - makes "FF default-deny + enhance with an extension" strategy look pretty solid now, although it is a bit of work

**Comment by earthlng** [comment]
^^ thanks Pants. I hope they use this opportunity to get rid of some of the filters that prevents extensions from seeing (+being able to modify or block) certain requests. For example I think it sucks that extensions cannot see requests made by other extensions.

**Comment by atomGit** [comment]
ok, off-topic by a 1/2 mile, but ima gonna PIMP MY SITE!!! - i really like my [smooth scrolling tweak](http://12bytes.org/tech/firefox/firefox-tweaks-and-fixes-and-things#enable-silky-smooth-scrolling) without an extension

**Comment by Thorin-Oakenpants** [comment]
FF59: auto canvas spoofing to 10x10 white canvas via RFP with prompt for "user initiated" canvas, go to Panopticlick (no site permission overrides for canvas - its is set as default, ask)

- Canvas Blocker set to "fake readout API"
   - `a273d6a847f0e2a57fa0161158f12fed` (one in 19.61 browsers)
- Canvas Blocker disabled
   - `a273d6a847f0e2a57fa0161158f12fed` (bingo! - matches when using fake readout API)
- Canvas Blocker set to block
   - `undetermined` (one in 38 browsers)
   - this was what I was using for FF58 so I didn't get any canvas prompt fatigue

Change site canvas permission to allow, CB on "fake readout API", retest several times
- `c3bdbe451d504911ac1db4494567fc6a`
- `5599703e36affa6d76d6779c165e1a01`
- every test is a fake new random unique hash

So ends the test for using RFP & CB (in fake readout API) together. If you happen to see a prompt, cool. If you allow it thru even by mistake, it's random faked. If you really need the real canvas, then allow in site panel and whitelist in CB.

I believe this is the best solution: always look like other RFP users, but use CB spoofery when needed

@kkapsner Thanks man :)

**Comment by Thorin-Oakenpants** [comment]
for the referers section

OK, so I looked it up - [uMatrix](https://github.com/gorhill/uMatrix/wiki/Per-scope-switches#spoof-referer-header)

> ... the referer information will now be foiled using the root URL derived from the URL of the request. For example, if the URL of a request is `http://www.example.com/blahblahblah/boring.html` and the referer is `http://google.com`, the referer will be spoofed using the `http://www.example.com/` string.

**Comment by mazesy** [comment]
I've had a tough time dealing with an issue related to [Firefox's Gesturefy extension](https://github.com/Robbendebiene/Gesturefy/issues/271)

It appears that 

```
/* 2516: disable PointerEvents
 * [1] https://developer.mozilla.org/en-US/docs/Web/API/PointerEvent ***/
user_pref("dom.w3c_pointer_events.enabled", false);
```

is the culprit. Resetting `dom.w3c_pointer_events.enabled` to true resolved the issue.

Remains: what is the importance of this `dom.w3c_pointer_events.enabled` setting?

**Comment by Thorin-Oakenpants** [comment]
^^ did you read the linked document - also https://github.com/ghacksuserjs/ghacks-user.js/issues/349#issuecomment-373723466 . It was false until FF59. Obviously Gesturefy leverages it from 59+ - maybe the dev could detect it is disabled and revert to pre 59+ methods.

**Comment by mazesy** [comment]
I hadn't read the description/comment #349, thanks for letting me know (I just cannot follow in real-time all the comments).
I read Gesturefy's dev answer. I'm not speculating in any way, my post above was only to inform all of us that at least one webextension issue was related to this setting, implicitly proposing that it be commented out.
There are so many settings in ghacks-user.js that inevitably a conflict may occur, which is also why this sticky thread is aimed at. In my case this is the first webextension conflict I encounter.

**Comment by Thorin-Oakenpants** [comment]
> which is also why this sticky thread is aimed at

Well, TBH, I envisaged this sticky as insights into how various prefs and extensions can enhance or conflict one another in regards to *various aspects of privacy/tracking/FP/security* etc - i mean look at OP

I don't really care if `dom.allow_cut_copy` breaks `Copy on Select`, or if `dom.w3c_pointer_events.enabled` breaks `Gesturefy` - as these extensions have nothing to do with security/privacy. Don't get me wrong, I don't `care` is not the same as not wanting to be told. I'm probably not going to bother adding anything to the user.js as it sets a precedence

**Comment by mazesy** [comment]
> I'm probably not going to bother adding anything to the user.js as it sets a precedence

That, I really don't understand, Pants. Users of Ghacks-user.js rely on settings which may have an impact on Firefox's usability (webextensions included) to be commented, they consider that all commented-out settings are free of any other impact than privacy and security. If it appears that a setting will conflict with a webextension I don't see why it wouldn't be noted. Understandable that all conflicts may not be guessed but once reported why not take them into consideration, leaving the user to his choice? Issue with Gesturefy 1.2.1 is new, was unknown, couldn't be known. Now that it is why not help other users to not lose precious time in finding out why a webextension doesn't run correctly once he's applied all non-commented settings?

Moreover, Pants, if you stick on your decision than why should I or anyone report an issue as I did?

**Comment by Thorin-Oakenpants** [comment]
> which may have an impact on Firefox's usability

Correct. On **FF's** usability, not extensions that add nothing to privacy/security/etc. Otherwise I would end up listing thousands of extensions

**Comment by Thorin-Oakenpants** [comment]
Because I commented, I am notified of any more comments etc on that Gesturefy issue. If it is resolved by them, cool, if not, I can revisit. Gesturefy has over 100k users, so that's another factor I would take into account. Note that there are currently zero extensions (non-privacy-etc ones) mentioned in the user.js (I can;t see any), so this would be a first, and believe me, it would open the floodgates

Drink some wine and go smoke your weed pipe for a bit :grin:

**Comment by Thorin-Oakenpants** [comment]
FYI: here is the bugzilla for RFP to deal with pointer events - https://bugzilla.mozilla.org/show_bug.cgi?id=1363508

**Comment by crssi** [comment]
Hey @Thorin-Oakenpants (and @mazesy).
Maybe we could have one a "non-debatable, only final findings" sticky issue with topic "breakages" where peoples could post culprits and working preferences workarounds.
"Non-debatable" I mean to avoid long topics, which would just clutter the topic. For a debate there should be separate issue topic.
This way, we could still avoid and precedence and mess in the original user.js and in the same time we can still inform users of that sorts of culprits.
I could immediately fill some in it.

What do you think?
Cheers

**Comment by Thorin-Oakenpants** [comment]
Here - https://github.com/ghacksuserjs/ghacks-user.js/issues/391 - have at it. Keep it clean and short and on topic or I will ditch it

**Comment by verlain3** [comment]
Correct if i'm wrong but shouldn't Cookie AutoDelete work with FPI already in Firefox 59?

**Comment by Thorin-Oakenpants** [comment]
Yup. Noted ages ago in the wiki and FPI sticky and even in the user.js. I don't use CA-D and some people still moan about FPI with it. My advice if turning FPI on, is to wipe all persistent data (ctrl-shift-del) time range everything, and everything checked as a one off cleaning, and then start again, because the OA have changed, so they're essentially all new distinct items.

**Comment by verlain3** [comment]
Is there an easy way to find if it's working as intended, i mean, if it's deleting localStorage, etc with CA-D and FPI turned on? Sorry if i'm being a bit off-topic.

**Comment by Thorin-Oakenpants** [comment]
I'm not 100% sure what CA-D deletes: the cookies for sure, localStorage I think was enabled in 58?, but IndexedDB cannot be cleared by host I believe (but you can clear all that on close at this stage = part of `Offsite Web Data`)

Note: unless you allow a site to set a cookie, you cannot get any localStorage or IndexedDB. Does CA-D have a blackisting, if so, then these sites you shouldn't need to worry about regards CA-D cleaning

Lets say you auto-delete cookies (and localStorage) when you leave github (I know github uses localStorage), then to test it
1. you would go to github and while staying on the page, hit Shift-F9 to bring up the Storage Inspector, and you can see you have localStorage and cookies items
2. Close github, wait for your x seconds for CA-D to do its thing, then check if the data is still there
   * load chrome://browser/content/preferences/cookies.xul in the urlbar and see if you have a cookie
   * load profile\webappsstore.sqlite and view the webappstore2 table for github entries

I would ask at CA-D for a better solution. I am sure there are scripts or dev tools that show the Origin Attributes etc, but that's not my world

**Comment by verlain3** [comment]
Thanks so much @Thorin-Oakenpants

**Comment by Thorin-Oakenpants** [comment]
> Note: unless you allow a site to set a cookie, you cannot get any localStorage or IndexedDB

to clarify: You can still get IDB entries set by service workers

**Comment by Thorin-Oakenpants** [comment]
@vertigo220 @crssi etc - updated OP to include the Referer section - comments welcome

**Comment by crssi** [comment]
Sweet :heart:

**Comment by Thorin-Oakenpants** [comment]
Cool .. got the crssi seal of approval :kiss:

**Comment by Thorin-Oakenpants** [comment]
updated `cookies [1]` added the info from https://github.com/gorhill/uMatrix/wiki/Cookies to clarify that uM deals with HTTP Cookie headers and that JS can still read that info (and thus send it out)

**Comment by vertigo220** [comment]
> if FF prefs allow web workers, but uMatrix blocks them, there may be instances where the web site detects workers are available but can't actually use them, leading to breakage, rather than falling back to the non-web-worker version [read this somewhere on uM repo but can't find it now]

https://www.reddit.com/r/uMatrix/comments/7p7adg/web_workers/

**Comment by Thorin-Oakenpants** [comment]
> it breaks the AMO site

That's no reason to disable it (permanently). Are you in there 24/7 installing extensions? Anyway, it's just the AMO site thinking you're a different version and changing what it displays accordingly. Its been known about for ages see [bugzilla 1394448](https://bugzilla.mozilla.org/show_bug.cgi?id=1394448)

Solution: internally code is fine, so if the extension is compatible with your real FF version, then just drag and drop the `Add to Firefox` button into your urlbar

> it changes your time zone

That's a feature, not a bug. The only exceptions will be for extension when they fix that - the idea being that all extensions be exempt from RFP

**Comment by Thorin-Oakenpants** [comment]
re: time zone

Either you buy into the full RFP package, and everyone spoofs the same timezone, or its pointless (there is talk of allowing site exceptions but I doubt it will happen). It's all `relative` anyway (that's a reference to Einstein). The times of posts etc in forums (and lets face it, people could post from any of 24 time zones) aren't inaccurate at all, they're still relative to each other. Seems like a server / website issue to me. Posts should use server time.

The only issues/comments I have seen are
- you need to be careful with looking up times for events, as it may give you a time for UTC 0
- gmail: eg your gmail app on your phone is x hrs ahead or behind UTC 0, but when using gmail in FF with RFP, you are seemingly sending emails from the future or replying from the past.

**Comment by vertigo220** [comment]
I'm not saying I plan to disable it permanently because of it, but I've had it disabled the past couple weeks while testing stuff, installing and trying out addons, and a feature like this should work for non-advanced users, who aren't going to know why the AMO site isn't working properly and know how to fix it. Not to mention you'd think Mozilla would make their feature not interfere with their own website. Just seems like a pretty significant oversight. Anyways, I was hoping someone knew of a pref or addon that would disable it on certain sites and/or AMO or that could override it so the user agent could be forced on those sites. Also, this should be mentioned in the OP, so people aren't caught off guard when they enable this pref then find AMO doesn't work right.

And I realize the time zone change is a feature, not a bug; I wasn't saying otherwise. I'm just saying that I would like the ability to have the benefits of resistFingerprinting without the subsequent issues caused by that. Ideally, it should provide a way to tell it to leave that alone. Since that's clearly not the case, again I was wondering if anyone knew of a way to override it.

**Comment by vertigo220** [comment]
> Either you buy into the full RFP package, and everyone spoofs the same timezone, or its pointless

I understand how fingerprinting and uniqueness works, and that I would be setting myself apart by changing my time zone, but I suspect I would still be much less unique than if not using the pref at all, and therefore that it would by no means be pointless.

**Comment by Thorin-Oakenpants** [comment]
> and a feature like this should work for non-advanced users

Well, that's why it's not been exposed to the public yet: no UI, no PB Mode testing done yet, no marketing/advertising, no shield studies yet. etc. It's not even completed - eg fonts FP'ng, and fixing RFP breakage.

> this should be mentioned in the OP

I don't follow. What does RFP have to do with this issue? Unless you meant that I provide some alternative for RFP

> again I was wondering if anyone knew of a way to override it.

I am not an expert, but an extension should do this since it can modify the data/headers on their way out (i.e after FF returns the result). There are loads of UA extensions, and I think something like Safe Script might have a time zone option.

**Comment by vertigo220** [comment]
> Well, that's why it's not been exposed to the public yet

I get that. All I was looking for was a workaround that hopefully somebody had found to fix this issue.

> I don't follow. What does RFP have to do with this issue? Unless you meant that I provide some alternative for RFP

I was saying that in your canvas section, where you mention to use the pref, a simple caution that it will cause AMO to not work properly could be added so anybody that reads it will know about the breakage, so then when they go to AMO they'll be aware of it, vs just enabling it and then going there and not knowing why it's not working. I'm not saying they should be told to use it with caution, just that they should be cautioned (or perhaps a better word would be informed) about the issue. That's all.

**Comment by Thorin-Oakenpants** [comment]
The repo is about the user.js, and we have RFP on. Everything RFP does is in the user.js. This sticky is about alternatives. RFP (currently) does a hell of lot more breakage than just AMO. I don't see the need to replicate info in more than one place if I can help it (it just creates more work).

The AMO issue is mostly moot now with FF60 (since the spoof is finally off 52). That's not to say that extensions can't specify a minimum version and the issue can still arise. You have your AMO solution (drag the `Add to Firefox` button to the urlbar). Time zone spoof, as I already said, maybe an extension can override it - I've never tried.

**Comment by Thorin-Oakenpants** [comment]
^ I added something to the OP .. see if you can spot it :)

**Comment by Thorin-Oakenpants** [comment]
here's someone with their time zone screwed up - https://old.reddit.com/r/firefox/comments/8iqlgl/time_zone_is_wrong/ - and it's not RFP (or OS). Maybe the thread will reveal the answer to be an extension

**Comment by Thorin-Oakenpants** [comment]
@vertigo220 I have not used this (but the dev seems legit, I'm testing out his AudioContext FP Defender) - https://addons.mozilla.org/en-US/firefox/addon/change-timezone-time-shift/ . If it had a whitelist, that would be quite cool (i.e only apply the new timezone to the whitelist). I do not recommend deviating from RFP though. Either use a different profile/browser for web mail, or learn to live with it

**Comment by crssi** [comment]
[CanvasBlocker](https://addons.mozilla.org/en-US/firefox/addon/canvasblocker/) 0.4.6-Alpha1 does now also Audio... see here; https://github.com/kkapsner/CanvasBlocker/issues/71#issuecomment-397759582

I am sure that @kkapsner would appreciate some testing from you guys too.

Cheers

**Comment by alexander255** [comment]
@Thorin-Oakenpants: I see you already updated the code URL. Now that @meh has given me full access rights on AMO, I've also create the privacy policy you requested a year ago: https://addons.mozilla.org/de/firefox/addon/smart-referer/privacy/
Sorry for the long delay, but I simply wasn't able to do this before today. :slightly_smiling_face:

**Comment by Kraxys** [comment]
Just discovered the FF addon  [Site Bleacher](https://addons.mozilla.org/en-US/firefox/addon/site-bleacher/). It automatically clears coolies, local storage **and indexedDB** as soon as the les tab relative to a domain is closed. Far less options than CAD but Site Bleacher clears indexedDB, what CAD can't.

**Comment by Thorin-Oakenpants** [comment]
there is no extension API available to clear IDB by domain

> and Indexed DBs are removed when first site tab is open, so sites are unable to access data from previous session

if I read that correctly, it says IDBs (plural) are removed when you visit your first website after opening the browser - see the words "from previous session". I don't think English is the author's first language - that sentence is very unclear.

You can already do this, btw, by clearing data on close

**Comment by crssi** [comment]
> if I read that correctly, it says IDBs (plural) are removed when you visit your first website after opening the browser

I read it as when fist tab of specific domain is opened the IDB is cleared, which makes sense

> I don't think English is the author's first language

If you look into his account, you will see he is from Serbia, Belgrade, so the English is not the first language

Cheers

**Comment by Thorin-Oakenpants** [comment]
Even FF internally can't clear IDB by domain... cogito ergo sum. Why don't you test it.

**Comment by crssi** [comment]
Well, yes and no. If WE injects a code to delete IDB into the visited site, then it can be deleted.
I assume that the "Site Bleacher" does that and deletes IDB if this is the first tab to open that particular site.

> Why don't you test it.

I have tested the Site Bleacher... not that I need it, since I am using TC for that purpose, but Site Bleacher is successful in deleting IDB.

So... do you test it?

Cheers

**Comment by atomGit** [comment]
`browser.display.use_document_fonts`

value can be toggled with [Toggle Fonts](https://github.com/M-Reimer/togglefonts) add-on ([AMO](https://addons.mozilla.org/en-US/firefox/addon/togglefonts/))

note that Toggle Fonts does not alter `gfx.downloadable_fonts.enabled`

**Comment by ilikenwf** [comment]
Re: canvas/canvasblocker and RFP...I have this PR in to waterfox. I haven't bothered trying to submit to Mozilla as they've said they don't want to divorce canvas functionality from RFP:

https://github.com/MrAlex94/Waterfox/pull/1559

**Comment by Thorin-Oakenpants** [comment]
You don't seem to understand fingerprinting very well. RFP & canvas currently lowers the entropy in that metric, in fact, it eliminates it by making everyone the same. As soon as you diverge from other RFP users, you stand out. Obviously it depends on the script, but I'm talking theoretically - which is how you need to approach it. Any hole is a hole that can be exploited.

It's not hard to detect RFP: you can just check timing rounding (which *could* be changed by a pref value - it sucks though because the check takes time: you could probably get away with 30ms). And there are other cumulative checks that could give it away (but not totally infallible). But since FF78 there's an even quicker method (less than 1ms, 100% correct, all the time).

And RFP does allow a site exception for canvas: it's not like you don't have some control.

Of the at least 8 methods to confound FPing, one is to lower entropy, one is to raise. Both have pros and cons. AFAIK, RFP canvas is going to become randomized - but until it does, don't undermine it - that's just stupid

---

Also: just so people are aware - don't use Waterfox: there is nothing it does that Firefox can't (there might be the one odd thing here or there: I'm not wasting my time checking). This does not include legacy extension support: which I consider the lack of in Firefox as a good thing, and the inclusion of in Waterfox as a security hole.
- https://github.com/ghacksuserjs/ghacks-user.js/issues/517#issuecomment-428039991

> I feel many many more users would switch to Waterfox, especially the ghacks user.js crowd

Not if I can help it. There is no reason to use a fork.

**Comment by ilikenwf** [comment]
I do understand and there are more than one school of thought on fingerprinting than just "blend in," although in this case, I blend in MORE using a faked fingerprint canvas hash than using the one generated by RFP according to panopticlick. 

The fact that all my other fingerprint attributes are inline with basically all *fox browsers based on ESR with a randomized fingerprint (as opposed to a 1 in 1200-1800) makes me less trackable using temp containers, as every container, every domain in every container gets it's own random fingerprint hash. 

Detection of RFP and canvas fakery doesn't matter, either, since the standard now is to lie to the webserver. How you lie is where people debate - do you want to blend in, or do you randomize your UA, canvas, and every other aspect, or mix and match. Free software is about freedom, stop trying to imply that we must trust the authority on this somehow.

Regardless, I choose to and will continue to  use Waterfox, because it doesn't prioritize politics and monetization like Mozilla does. Philosophy matters with software, not just the code itself. 

Even Archlinux includes a Google tracking id in their Firefox builds...for one...and the lack of pocket and other complete and other bullsh*t inclusions that Mozilla has pushed on users makes Waterfox much more attractive. I also remember Mozilla installing an addon silently to market some TV show...showing that not only can they do that without user interaction, but will do so proves that they are no longer trustworthy without heavy auditing.

**Comment by Thorin-Oakenpants** [comment]
OK, you really **don't** understanding FPing

**Comment by B00ze64** [comment]
Just visited uMatrix repos, GH has Archived it all, says he won't spend any more time on uMatrix. uBO is nice and all for cosmetic filtering but it does not come anywhere close to the control you have in uMatrix. Disappointing - no discussion, no warning, just a guy opening an issue and getting told "forget it I'm done" and poof, everything archived :-(

**Comment by atomGit** [comment]
not surprising at all given the number of commits on uBO vs. uM

i'm very disappointed by this - there's lots i could say about this but i'll refrain, suffice to say that i think some of his decisions are idiotic, starting with splitting uBO off from uM

**Comment by Xenoslyce** [comment]
What should be done to make uBlock Origin handle trackers instead of the browser? I use AdNausuem so it'd be nice to be able to send a message
